File: solutions/Leetcode_32/leetcode_32.py
class Solution1:
    def longest_valid_parentheses(self, s: str) -> int:
        stack = []
        prev = 0
        max_length = 0
        for x in s:
            if x == '(':
                stack.append(prev)
                prev = 0
            elif len(stack) > 0:
                prev = prev + stack.pop() + 2
                max_length = max(max_length, prev)
            else:
                prev = 0
        return max_length


class Solution2:
    def longest_valid_parentheses(self, s):
        dp = [0] * (len(s)+1)
        max_len = 0
        for i in range(1, len(s)):
            if s[i] == ')':
                if s[i-1] == '(':
                    dp[i] = dp[i-2] + 2 if i-2 >= 0 else 2
                elif i - dp[i-1] - 1 >= 0 and s[i-dp[i-1]-1] == '(':
                    dp[i] = dp[i-1] + 2 + dp[max(i-dp[i-1]-2, 0)]
                max_len = max(max_len, dp[i])
        return max_len


if __name__ == '__main__':
    s = ")()())()()("
    ans = Solution2().longest_valid_parentheses(s)
    print(ans)

File: solutions/Leetcode_32/leetcode_32.cpp
#include <iostream>
#include <string>
#include <stack>
#include <vector>

using namespace std;

class Solution1
{
public:
    int longestValidParentheses(string s)
    {
        stack<int> stk;
        int prev = 0, max_length = 0;
        for (auto x : s)
        {
            if (x == '(')
            {
                stk.push(prev);
                prev = 0;
            }
            else if (!stk.empty())
            {
                prev = prev + stk.top() + 2;
                max_length = max(max_length, prev);
                stk.pop();
            }
            else
            {
                prev = 0;
            }
        }
        return max_length;
    }
};

class Solution2
{
public:
    int longestValidParentheses(string s)
    {
        vector<int> dp(s.length() + 1, 0);
        int max_len = 0;
        for (int i = 1; i < s.length(); i++)
        {
            if (s[i] == ')')
            {
                if (s[i - 1] == '(')
                    dp[i] = (i - 2) >= 0 ? (dp[i - 2] + 2) : 2;
                else if (i - dp[i - 1] - 1 >= 0 && s[i - dp[i - 1] - 1] == '(')
                    dp[i] = dp[i - 1] + 2 + ((i - dp[i - 1] - 2 >= 0) ? dp[i - dp[i - 1] - 2] : 0);
                max_len = max(max_len, dp[i]);
            }
        }
        return max_len;
    }
};

int main()
{
    string s = "(()())";
    int ans = Solution2().longestValidParentheses(s);
    cout << ans << endl;
}